SQL FS

jforissier opened the pull request:
Depends on https://github.com/OP-TEE/optee_client/pull/50.

jbech-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,878 @@
+	return (raw_pos - meta_size()) / block_size_raw();
+}
+
+/* Retur the position of a block in the DB file */
```
Nitpick: Retur -> Return

jbech-linaro commented:
I've done a quick review Jerome, but I'd like to discuss with you so we're on the same page. I've added this as an item for our 1:1 next week. After that I'll do a more thorough review.

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	paddr_t phpayload = 0;
+	uint64_t cpayload = 0;
+	struct tee_fs_rpc *bf;
+	int res = -1;
```
Can we define non-magic return values in this function?

jforissier replied:
Yes, this function should probably return TEE_Result. I'll fix it.

Regarding the other functions, they simply return the status value set by tee-supplicant, unless communication failed in which case they return -1. The convention is: `<0` means failure (in this case the actual value is system-specific, can be -(errno) as explained in [1]), `>=0` is success and depending on the function it may be a useful value like a file descriptor or a number of bytes.

So... I propose to `#define RPC_FAILED -1` and use that instead of the hardcoded  `-1` to initalize `int rc = -1;` in the below functions (`tee_fs_rpc_read` and so on). And, to add the above explanation as a comment in `tee_fs_rpc.h`. Does it sound OK?

[1] https://github.com/OP-TEE/optee_client/pull/50/files#diff-4b1fc55b89725a5c4b5b364946cc554eR412

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	assert(cmd_id == OPTEE_MSG_RPC_CMD_FS ||
+	       cmd_id == OPTEE_MSG_RPC_CMD_SQL_FS);
+
+	thread_rpc_alloc_payload(sizeof(struct tee_fs_rpc) + len,
```
Should we check if `len == 0` before calling this? I.e, just return (either OK or with an error).

jforissier replied:
`len == 0` is a valid case, it means there is no additional data to be appended to the `struct tee_fs_rpc`. See for instance `tee_fs_rpc_close`.

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	int rc = -1;
+	size_t len;
+
+	DMSG("(%d, %s, %d)...", id, name, mode);
```
Messages like this and further (in other functions also) down for return codes are useful, however, since the print doesn't tell what the number etc it represent, it makes this less useful. I'd change the message slightly.
`DMSG("id: %d, name: %s, mode: %s");`
for example.

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+
+	head.op = TEE_FS_READ;
+	head.fd = fd;
+	head.len = (uint32_t) len;
```
Nit-pick, No space between the cast and len.

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+
+	head.op = TEE_FS_WRITE;
+	head.fd = fd;
+	head.len = (uint32_t) len;
```
Same nit-pick here (space ...)

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	if (!name)
+		goto exit;
+
+	len = strlen(name) + 1;
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+		goto exit;
+
+	len = strlen(name) + 1;
+
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+		return -1;
+
+	len = strlen(path) + 1;
+
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	DMSG("(%d, %s, %d)...", id, file, flags);
+
+	len = strlen(file) + 1;
+
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+		goto exit;
+
+	len = strlen(name) + 1;
+
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	DMSG("(%d, %s)...", id, name);
+
+	len = strlen(name) + 1;
+
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,572 @@
+	int rc = -1;
+
+	DMSG("(%d, %s)...", id, file);
+
```
```
if (len <= 1)
    goto exit;
```
That save one RPC call ..

jbech-linaro commented:
This is all good and even though there are a little bit more code now, it seems cleaner in general. However, we're still doing the mistake that we're still implementing a POSIX like API on secure side. One of the key goals with the FS simplifications was to try to get rid of that interface and instead try to do the minimum amount of work on secure side and tunnel as much as possible to tee-supplicant. Without having looked that deeply into it, it still might be that we just need a read and a write call (almost like `tee_fs_rpc_send_cmd()`)

As of now, I think we could go on merging this patch (and its counterpart in optee_client). But as the next step and as soon as possible, we should try to look at the GP Internal spec Trusted Storage section and come up with what needs to be kept on secure side and what could be handled by tee-supplicant.

Along with that work we also need to identify functionality that needs to be handled differently (uniquely) for the different file-system types. For example for RPMB we need to do some pre-processing (MAC'ing) before sending the data to secure side. That's a special case for RPMB. Maybe we can add some hooks for doing things based on the storage type.

jforissier commented:
@jbech-linaro thanks for reviewing. Hopefully I have addressed all your comments.
Now regarding further simplification by moving more code to normal world: I would surely like to do so, but it's not obvious to me how it can be done.

jbech-linaro commented:
@jforissier , understood. Let's try to dig into it to see what we can come up with (regarding FS simplifications).

Thanks for fixing so promptly. Feel free to add my:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

jbech-linaro commented:
By the way, I think we should wait merging this until after we've release 2.1.0. I.e, my suggestion is to let's try to not make many more big changes to OP-TEE this week.

jforissier commented:
@jbech-linaro no problem :)
And, I should probably wait for more reviews before I rebase/squash.

etiennec-linaro commented on core/include/tee/tee_fs_rpc.h:
```diff
@@ -0,0 +1,102 @@
+/*
+ * Return values:
+ *   < 0: error. The actual value is meaningless (see below).
+ *  >= 0: success. The value may be a file descriptor, a number of byte, or
```
minor: "a number of bytes"

jforissier replied:
OK will fix

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,600 @@
+	if (!ALIGNMENT_IS_OK(phpayload, struct tee_fs_rpc))
+		goto exit;
+
+	bf = phys_to_virt(phpayload, MEM_AREA_NSEC_SHM);
```
you should check that payload fully fits inside SHM, not just the start address.

jforissier replied:
Sorry I don't get it. If `thread_rpc_alloc_payload()` returns a non-NULL value, it means we were able to allocate `sizeof(struct tee_fs_rpc) + len` bytes inside the shared mem, doesn't it? So what exactly should be checked you think?

etiennec-linaro replied:
i didn't see `thread_rpc_alloc_payload()` checking that allocated buffers really are inside the SHM.
if return buffer is physically at end of SHM, accessing it could overflow to some secure (or not) memory.

jforissier replied:
I still don't get it. Care to share the code you would add here?

jenswi-linaro replied:
I think `thread_rpc_alloc_payload()` should be responsible for checking that the returned memory is OK to use or we would duplicate code everywhere `thread_rpc_alloc_payload()` is called.

etiennec-linaro replied:
> I think thread_rpc_alloc_payload() should be responsible for

> checking that the returned memory is OK to use...

agree

On 20 July 2016 at 13:42, Jens Wiklander <notifications@github.com> wrote:

> In core/tee/tee_fs_rpc.c
> <https://github.com/OP-TEE/optee_os/pull/797#discussion_r71509045>:
>
> > +	uint64_t cpayload = 0;
> > +	struct tee_fs_rpc *bf;
> > +	int res = TEE_ERROR_GENERIC;
> > +
> > +	assert(cmd_id == OPTEE_MSG_RPC_CMD_FS ||
> > +	       cmd_id == OPTEE_MSG_RPC_CMD_SQL_FS);
> > +
> > +	thread_rpc_alloc_payload(sizeof(struct tee_fs_rpc) + len,
> > +				 &phpayload, &cpayload);
> > +	if (!phpayload)
> > +		return TEE_ERROR_OUT_OF_MEMORY;
> > +
> > +	if (!ALIGNMENT_IS_OK(phpayload, struct tee_fs_rpc))
> > +		goto exit;
> > +
> > +	bf = phys_to_virt(phpayload, MEM_AREA_NSEC_SHM);
>
> I think thread_rpc_alloc_payload() should be responsible for checking
> that the returned memory is OK to use or we would duplicate code everywhere
> thread_rpc_alloc_payload() is called.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/797/files/772653438970a01729cab6910e1776e840697056#r71509045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEHJl5gUeHn6Dg7ZTP8iadlpnIKgfKLks5qXgm5gaJpZM4IhS8P>
> .
>

jenswi-linaro replied:
Taken care of in https://github.com/OP-TEE/optee_os/pull/930

jforissier replied:
@jenswi-linaro thanks. In fact I was so persuaded that `thread_rpc_alloc_payload()` was already checking this that I did not bother to double check! Hence the misunderstanding.

jenswi-linaro replied:
Yeah, I was a bit surprised myself. :-)

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,600 @@
+	char *tmp = NULL;
+	TEE_Result res;
+	int rc = RPC_FAILED;
+
```
why not checking "old, new" whereas such are tested in `tee_fs_rpc_mkdir`,`tee_fs_rpc_open`

jforissier replied:
You're right, there is not reason why they should not be checked. Will fix.

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,600 @@
+
+int tee_fs_rpc_rename_rpc(int id, const char *old, const char *nw)
+{
+	size_t len_old = strlen(old) + 1;
```
same comment as above: exit with error if `old` or `nw` are NULL ?

jforissier replied:
yep!

matt2048 commented on documentation/secure_storage.md:
```diff
@@ -10,14 +10,16 @@ integrity of the data stored and the atomicity of the operations that modifies
+enabled by `CFG_SQL_FS=y`. See [secure_storage_sql.md](secure_storage_sql.db).
 
 It is possible to use the normal world filesystem and the RPMB implementations
 simultaneously. For this, two OP-TEE specific storage identifiers have been
```
minor, could we also use sql fs  as  the former two simultaneously, there should be three storage identifiers

jforissier replied:
Yes it is possible, we do have 3 identifiers, I just need to rephrase the text that follows my addition. Good catch, will fix.

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,600 @@
+	return rc;
+}
+
+int tee_fs_rpc_open(int id, TEE_Result *errno, const char *file, int flags)
```
`errno` seems quite useless. why not drop it as for other APIs.

jforissier replied:
Agreed.

etiennec-linaro commented on core/tee/tee_ree_fs.c:
```diff
@@ -245,339 +162,79 @@ static int tee_fs_send_cmd(struct tee_fs_rpc *bf_cmd, void *data, uint32_t len,
  */
 #define REE_FS_NAME_MAX (TEE_FS_NAME_MAX + 20)
 
```
as all routnies below only call the generic `tee_fs_rpc_xxx(OPTEE_MSG_RPC_CMD_FS, ...`, could local caller call it straight instead of `ree_fs_rpc_xxx(...)`. It seems that there will not be anything 'ree fs' specific to add inside these routines.

jforissier replied:
That's correct. I can simplify this (and similarly in `tee_sql_fs.c`).

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,879 @@
+
+exit:
+	free(ct);
+	return rc;
```
for the same `rc` value, `errno` will or will not carry an info. 
is that ok ? caller is assumed to have pre-init `errno` ?

jforissier replied:
No, we'd better make sure `errno` is set to something different from `TEE_SUCCESS` each time `rc` is `< 0`. I think it's OK to initialize it to `TEE_ERROR_GENERIC` on entry, as it is done in `read_meta()`. Because the caller is supposed to check `errno` only in case `rc < 0`.

etiennec-linaro replied:
ok
check also in `create_meta()` and `read_meta()`. Seems this only suspicious places here.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,879 @@
+		       out_size);
+#endif
+	} else {
+		/* Unexpected data length */
```
shouldn't this trig some error ? `rc = ....`

jforissier replied:
Yes, will do `rc = -1`.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,879 @@
+	fdp->flags = flags;
+
+	fd = read_meta(errno, fdp);
+	if (fd < 0)
```
`read_meta` can return several positive values. Seems only rc=meta size if a valid one in case operation succeed.
If so, it should be checked (maybe from `read_meta()`, returning positive or null only for expected meta sizes read.

jforissier replied:
`read_meta` will return 0 the first time the file is opened (and the meta-data header is created by `create_meta()`.

etiennec-linaro replied:
if meta-datas need to be created, `create_meta()` calls `write_meta()` which exits with `rc < 0` on error, and `rc = 0` on success as it checks expected nb of bytes were written successfully.

I think `read_meta()` should do the same:
return with < 0 in case of error
return with 0 if successful and meta-data have the expected size.

jforissier replied:
Yes this will work, I'll do as you suggest.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,879 @@
+	}
+
+	start_block_num = block_num(fdp->pos);
+	end_block_num = block_num(fdp->pos + len + 1);
```
if `(fdp->pos + len)` equals the start address of a block, `end_block_num = block_num(fdp->pos + len + 1);` will go 1 block ahead than expected.

jforissier replied:
Good catch, should be `fdp->pos + len - 1` obviously.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,879 @@
+	}
+
+	start_block_num = block_num(orig_pos);
+	end_block_num = block_num(fdp->pos - 1);
```
if don't catch this 
```
start_block_num = block_num(fdp->pos);
end_block_num = block_num(fdp->pos - 1);
```

jforissier replied:
Oops! That should be:
```
	start_block_num = block_num(fdp->pos);
	end_block_num = block_num(fdp->pos + len - 1);
```
Will fix.

etiennec-linaro commented on documentation/secure_storage_sql.md:
```diff
@@ -0,0 +1,77 @@
+
+## Introduction
+
+This document describes the SQL DB secure storage in OP-TEE, which is enabled
```
voc: rather "that" then "which"

jforissier replied:
Hmm... I don't think so: http://www.writersdigest.com/online-editor/which-vs-that
Maybe a native speaker can confirm or infirm?

d3zd3z replied:
"Which" is better. I'm not sure if being a native speaker really helps here. After all, we have a popular grammar book "Elements of Style" which is full of incorrect rules, including concerning which/that.

etiennec-linaro replied:
that for the link and the english lesson :)

matt2048 commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,1321 @@
+
+	if (!fdp) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
```
if go here, will go to exit point and there is a  "sql_fs_end_transaction_rpc(true)" call, but is this call needful and what is the result there will be after this call

jforissier replied:
I think the call will do nothing if not currently in a transaction. But you're right, this is not good style. I will change this to jump after `sql_fs_end_transaction_rpc()` instead. And same in other places.

matt2048 commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,1321 @@
+
+	if (!len) {
+		res = 0;
+		goto exit;
```
same as above "sql_fs_end_transaction_rpc(true)" call

matt2048 commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,1321 @@
+
+	if (!buf) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
```
also here as above.

matt2048 commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,1321 @@
+ * file in the SQL filesystem, and has the following structure:
+ *
+ * [       File meta-data       ][      Block #0        ][Block #1]...
+ * [meta_header|sql_fs_file_meta][block_header|user data][        ]...
```
through source code, I can see there is already encrypted_fek in meta_header struct, but sql_fs_file_meta struct in this file also has the encrypted_fek field, is it a duplication field when write meta-data to sql fs file?

jforissier replied:
Good catch. `encrypted_fek` should be moved to `struct sql_fs_fd` (the runtime descriptor for the file). I will fix this.

jforissier commented:
Review comments addressed.

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,605 @@
+
+	DMSG("(id: %d, old: %s, nw: %s)...", id, old, nw);
+
+	if (!old || !nw)
```
actually this can be useless as `old` and `nw` are already used above to init `len_old/new`.
could you move init of `len_old` and `len_new` below this check ?

jforissier replied:
Of course, sorry for the stupid mistake. I need to take more care when incorporating review comments.

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,605 @@
+	int rc = RPC_FAILED;
+	size_t len;
+
+	DMSG("(id: %d, name: %s, mode: %d)...", id, name, mode);
```
dummy question: why parenthesis and dots in the log message ?

jforissier replied:
The dots are intended to better differentiate the function entry and exit. The parentheses are meant to group the parameters between the function name and the dots. I found this format rather easy to read as I was debugging this code. Here is an example log:
```
DEBUG:   [0x0] TEE-CORE:sql_fs_write:792: (fd: 0, buf: 0x3f07cfd8, len: 36)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_begin_transaction:138: (id: 8)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_begin_transaction:151: ...0
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:264: (id: 8, fd: 0, offset: 52, whence: 1)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:277: ...52
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_read:390: (id: 8, fd: 0, buf: 0x3f051430, len: 4124)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_read:411: ...0
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:264: (id: 8, fd: 0, offset: 52, whence: 1)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:277: ...52
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:495: (id: 8, fd: 0, buf: 0x3f051430, len: 4124)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:515: ...4124
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:264: (id: 8, fd: 0, offset: 0, whence: 1)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_lseek:277: ...0
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:495: (id: 8, fd: 0, buf: 0x3f053420, len: 52)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_write:515: ...52
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_end_transaction:184: (id: 8, rollback: 0)...
DEBUG:   [0x0] TEE-CORE:tee_fs_rpc_end_transaction:196: ...0
DEBUG:   [0x0] TEE-CORE:sql_fs_write:851: ...36
```

d3zd3z replied:
The dots look a little like something got left out, but I'm not sure what else could be put there. I wonder if an arrow would look better. Either case, I'd probably have to look at the code to figure out what it meant, though.

jforissier replied:
What is left out is the return code, which is printed later. `foo()... ...0`. Arrows would be OK too.

etiennec-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -0,0 +1,605 @@
+	DMSG("(id: %d, old: %s, nw: %s)...", id, old, nw);
+
+	if (!old || !nw)
+		goto exit;
```
init `len_old` and `len_new` after `old` and `nw` are checked.

jforissier replied:
yup

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,871 @@
+
+	res = tee_fs_generate_fek(fdp->encrypted_fek, TEE_FS_KM_FEK_SIZE);
+	if (res != TEE_SUCCESS)
+		return -1;
```
should set `*errno = TEE_ERROR_GENERIC` (or `TEE_ERROR_SECURITY` ?) at routine entry.
(or at least in this error case)

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,871 @@
+	if (!fdp)
+		goto exit_ret;
+
+	*errno = TEE_ERROR_GENERIC;
```
from [`errno` man page](http://man7.org/linux/man-pages/man3/errno.3.html):
```
       Valid error numbers are all nonzero; errno is never set to zero by
       any system call or library function.
```
we should stick to this, no?

jforissier replied:
Yes, I agree that it's always better to stick to standards and well-known ways of doing things to avoid mistakes.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,871 @@
+	if (rc != (int)ct_size) {
+		rc = -1;
+	} else {
+		*errno = TEE_SUCCESS;
```
from [`errno` man page](http://man7.org/linux/man-pages/man3/errno.3.html):
```
       Valid error numbers are all nonzero; errno is never set to zero by
       any system call or library function.
```
we should stick to this, no?
could apply here and 2 other places of tee_sql_fs.c.

Actually, i would propose to change tee_sql_fs.c as following:
- default load `*errno = TEE_ERROR_GENERIC;` at entry of the exported sqlfs routines `sql_fs_open/read/write/lssek/ftruncate()`
- remove all `*errno = TEE_SUCCESS;`
- only modify `*errno = TEE_ERROR_xxx;` if we want to be more precise than `TEE_ERROR_GENERIC` on specific error (out of mem, ...)

jforissier replied:
Yes, and that would be consistent with initializing `rc` to `-1` in many places. Will do that.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,871 @@
+					  fdp->encrypted_fek);
+		if (res != TEE_SUCCESS) {
+			*errno = res;
+			goto exit;
```
This is an error case but this routine will returns with `rc` > 0 (`rc` is meta-data size) that will be considered valid by the caller.
Should force `rc = -1` or like.

jforissier replied:
Yes. Will fix.

etiennec-linaro commented on core/tee/tee_sql_fs.c:
```diff
@@ -0,0 +1,871 @@
+					  &ct_size, fdp->encrypted_fek);
+		if (res != TEE_SUCCESS) {
+			*errno = res;
+			goto exit;
```
should force `rc = -1`

jforissier replied:
Will fix.

jforissier commented:
Updated with comments addressed.

d3zd3z commented:
```
Reviewed-by: David Brown <david.brown@linaro.org>
```

jforissier commented:
@etienne-lms I think I have addressed all your comments, can you check please?

etienne-lms commented:
ok. sorry, it's been a while....

On 12 August 2016 at 15:33, Jérôme Forissier <notifications@github.com>
wrote:

> @etienne-lms <https://github.com/etienne-lms> I think I have addressed
> all your comments, can you check please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/797#issuecomment-239447148>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATEHJp8KI-NUCj8IyuF6oAuYaXHpVEtRks5qfHYngaJpZM4IhS8P>
> .
>

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
